### ISSUE TITLE
Question: Section heading level choice

### ISSUE THREAD

**Issue by zachcran** [LilSpazJoekp/docstrfmt#91]
`docstrfmt` uses the section heading characters `=` > `-` > `~` > `+` > `.` > `,`. What is your motivation for using these characters, as opposed to the section header characters that [Python suggests](https://devguide.python.org/documentation/markup/#sections) and [Sphinx recommends](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections) (`#` > `*` > `=` > `-` > `^` > `"`)?

I know it doesn't matter to the reST specification, but I'm curious since `docstrfmt` goes against what I know as the standard.

**Comment by LilSpazJoekp** [comment]
That was what it was in rstfmt and this project is a fork of it and I didn't think to change it.

**Comment by zachcran** [comment]
rstfmt seems to have updated their section headers to the standard set by Python and Sphinx that I mentioned above (see their latest [sample.rst](https://raw.githubusercontent.com/dzhu/rstfmt/master/sample.rst)). Would you be open to updating as well?

**Comment by LilSpazJoekp** [comment]
Absolutely! I'll get this updated as well!

**Comment by zachcran** [comment]
That is great, thank you! I really like the project so far btw!

**Comment by LilSpazJoekp** [comment]
That's fine with me if you want to make that change!

**Comment by anjos** [comment]
I'm looking at how to best address this from the perspective of this package, which focuses on formatting individual files. Despite Python having put forward a recommendation on the part/chapter/section organisation, projects may follow this guideline on individual files, or on a project level, across multiple sources (e.g. on Sphinx projects).  On those, for consistency, a part or chapter header may be set on the main file, while other (included) files only contain sections.

Moreover, the text documents may follow a standard, while the code docstrings may follow a different one (e.g. the one that is suggested by numpydoc/ruff instead).  Beyond this, there is a question about the use of only underlines and overlines for certain headers.  The official recommendation is to use overlines and underlines on the first 2 levels.

Given the variety of situations, perhaps a good compromise would be to do the following:

1. If the user does not explicitly define it, the formatter should just keep whatever is there
2. If the user signals (e.g. via a CLI flag), that the formatter should modify the headers, then docstrfmt uses the [values recommended by Python](https://devguide.python.org/documentation/markup/#sections) to do so, respecting overline/underline considerations to match the recommendation.
3. The CLI flag accepts an *optional* argument that defines which characters should be used for section headers, in which case the default is replaced by this input set.  A special marker separates headings that should be overlined and underlined from those that should only be underlined.

For example:

```python
# the pipe character marks the start of simple underline
# order taken from: https://devguide.python.org/documentation/markup/#sections
SECTION_CHARS = "#*|=-^\"'~+.`_:"
```

This would mark that the first 2 levels should be using "#" (hash) and "*" (asterisk) with both underline and overline, like so:

```rst
###########
 This is a part
###########

******************
 This is a chapter
******************

This is a section
============

This is a subsection
--------------------
```

This strategy would allow, if the user wishes so, multiple formats to exist within a single complex project (think documents vs. docstrings), while allowing the user to ensure consistency on sets that should be consistent.  

Is this a reasonable approach?

**Comment by anjos** [comment]
In what I have in mind, there won't be any "relative" formatting - if there are 2 different sets of header formats the user wants to have, then they would have to call docstrfmt twice, with different sets of parameters.  In this way, the behaviour remains predictable.

That said, we can later (ie. on another PR) think of having a more thorough configuration for a "single-pass" usage.  This could be done, e.g., by reading parameters from the `tool.docstrfmt` section of a `pyproject.toml` that specify which rules to apply to which files on the project.

**Comment by anjos** [comment]
Apparently, implementing point 1 (seemingly the simplest) requires we keep track of the current under/overline for each section, which does not seem to be trivial with docutils.  Any ideas welcome!

**Comment by LilSpazJoekp** [comment]
Yeah unfortunately some nodes lose info like that when parsed. Other than reading the file directly for the current character uses, I don't have any ideas to make that work.

**Comment by zachcran** [comment]
I like the proposed solution! I am only a user in this regard, though, so I do not have any suggestions to help with this either. It would be nice to have a stop-gap solution where the default section adornment characters were changed, though, to help maintain consistency with Python recommendations.

> Beyond this, there is a question about the use of only underlines and overlines for certain headers. The official recommendation is to use overlines and underlines on the first 2 levels.

@anjos Can you provide the source for the official recommendation you are mentioning here? I struggled to find it in the [reST specification](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#toc-entry-8), [Python reST guide](https://devguide.python.org/documentation/markup/#sections), or [Sphinx reST primer](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections). I could have easily missed it, and would like to have it as a reference.

While searching, I also found that apparently [there is a distinction](https://docutils.sourceforge.io/docs/user/rst/quickstart.html#toc-entry-6) between overline/underline and underline-only styles, even when they use the same character. Perhaps it is best to follow the user's lead on this and maintain the style contained in the document instead of prescribing a number of headers that will be converted to overline/underline. More often than not in the pages I am linking, I am seeing it explicitly stated that the overline is optional, as well as the inset title text, although they do include the overline and inset in many examples.

**Comment by anjos** [comment]
Source: https://devguide.python.org/documentation/markup/#sections (this was the one you already mentioned before).  I don't think there is any "official" source.

It reads: 

```
Normally, there are no heading levels assigned to certain characters as the structure is determined from the succession of headings. However, for the Python documentation, here is a suggested convention:

# with overline, for parts
* with overline, for chapters
=, for sections
-, for subsections
^, for subsubsections
", for paragraphs
```

> While searching, I also found that apparently [there is a distinction](https://docutils.sourceforge.io/docs/user/rst/quickstart.html#toc-entry-6) between overline/underline and underline-only styles, even when they use the same character. Perhaps it is best to follow the user's lead on this and maintain the style contained in the document instead of prescribing a number of headers that will be converted to overline/underline (if docstrfmt does not already do this). More often than not in the pages I am linking, I am seeing it explicitly stated that the overline is optional, as well as the inset title text, although they do include the overline and inset in many examples.

This is how it is implemented.

**Comment by LilSpazJoekp** [comment]
I would prefer docstrfmt to be like black in regard to deterministic behavior. For one, I would like to it to ignore any kind of inset the user has made on the headers. Does providing an overline change how it is rendered in rst? If not, then I don't see a reason in preserving this behavior.

**Comment by anjos** [comment]
> Does providing an overline change how it is rendered in rst? If not, then I don't see a reason in preserving this behavior.

Please clarify - would you like me to take action on the PR?

**Comment by LilSpazJoekp** [comment]
Yeah at a minimum it shouldn't inset the text. For overlines, since that wasn't historically preserved, it shouldn't start doing so now.

To be honest, I'm not sure I would add a parameter to configure section characters. The existing configurations are there to allow compatibility with other formatters. I just don't want to see people fussing over the character used when it doesn't change anything in the rendered result. However, I do agree that it should follow the recommended character order. 

Adding a bunch of configurations add more complexity and edge cases that could arise.

**Comment by anjos** [comment]
> Yeah at a minimum it shouldn't inset the text. For overlines, since that wasn't historically preserved, it shouldn't start doing so now.

OK.

> To be honest, I'm not sure I would add a parameter to configure section characters. The existing configurations are there to allow compatibility with other formatters. I just don't want to see people fussing over the character used when it doesn't change anything in the rendered result. However, I do agree that it should follow the recommended character order.
> 
> Adding a bunch of configurations add more complexity and edge cases that could arise.

The configuration is there so that the tool is not enforcing a particular behaviour on edge cases: I'm thinking particularly of docstrings in which all sections will become overlined headers.

If I understand this correctly, you are suggesting we should just change the default `SECTION_CHARS`, and everything else I added to the PR should go away?

**Comment by LilSpazJoekp** [comment]
I'm going to recant my previous ask as I don't want effort already expended to be wasted.

> If I understand this correctly, you are suggesting we should just change the default SECTION_CHARS, and everything else I added to the PR should go away?

No. I think the PR is good as is. I just want to check in here and see if there are any other changes you want to make to it.

I've done some more thinking on this and I do think this should be configurable since different linters enforce different styles and would cause docstrfmt and the other linter to fight with each other (it's why isort has a capability mode with black since they both can fight with each other).
